js/qz-tray.js

```diff
@@ -767,12 +767,67 @@ var qz = (function() {
                 }
                 // Promise won't reject on throw; yet better than 'undefined'
                 throw new Error("A connection to QZ has not been established yet");
-            }
+            },
+
+            uint8ArrayToHex: function(uint8) {
+                return Array.from(uint8)
+                    .map(function(i) { return i.toString(16).padStart(2, '0'); })
+                    .join('');
+            },
+
+            uint8ArrayToBase64: function(uint8) {
+                /**
+                 * Adapted from Egor Nepomnyaschih's code under MIT Licence (C) 2020
+                 * see https://gist.github.com/enepomnyaschih/72c423f727d395eeaa09697058238727
+                 */
+                var map = [
+                    "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U",
+                    "V", "W", "X", "Y", "Z", "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p",
+                    "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "+", "/"
+                ];
+
+                var result = '', i, l = uint8.length;
+                for (i = 2; i < l; i += 3) {
+                    result += map[uint8[i - 2] >> 2];
+                    result += map[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
+                    result += map[((uint8[i - 1] & 0x0F) << 2) | (uint8[i] >> 6)];
+                    result += map[uint8[i] & 0x3F];
+                }
+                if (i === l + 1) { // 1 octet yet to write
+                    result += map[uint8[i - 2] >> 2];
+                    result += map[(uint8[i - 2] & 0x03) << 4];
+                    result += "==";
+                }
+                if (i === l) { // 2 octets yet to write
+                    result += map[uint8[i - 2] >> 2];
+                    result += map[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
+                    result += map[(uint8[i - 1] & 0x0F) << 2];
+                    result += "=";
+                }
+                return result;
+            },
         },
 
         compatible: {
             /** Converts message format to a previous version's */
             data: function(printData) {
+                // special handling for Uint8Array
+                if (printData.constructor === Object && printData.data instanceof Uint8Array) {
```

Comment:
@akberenz can you give this a second review please?  In my testing `printData.constructor === Array`, not `Object`.

Reply:
Ok, it's because we wrap it into a double-array just before this:

https://github.com/qzind/tray/blob/94f02af092fbdd370f21672ba208f26342a1ae3a/js/qz-tray.js#L1570-L1576

We leverage a for-loop in the `relative` function, but fail to in the `compatible.data` function.